### PR TITLE
Cryptography checks written and tested

### PR DESCRIPTION
**Pull request by dhaugh**
This revision adds the crypto flaws plugin.

### CODE REVIEW

File: tests/test_plugins/test_crypto_plugins/keys/dsa-key.pub [hunk 1]
```diff
@@ -0,0 +1 @@
+ssh-dss AAAAB3NzaC1kc3MAAACBANTcOLxTjiR+AyAKJvsLNbJlXrfxBbZryKk5qacc9jEONOink4DdSsqJBwYBAD7RqF7luT3vc+8nyPzkxR9yG+OJ1WAxoBdJBs8ew184K8Ne0Jo6Sj0vkv2FSazp6f+l6tBkXdvyeIuwk4n9Ky+SwVCQCDyqJ5YKB+X4chvkzkzJAAAAFQCcRnwbjfb26nzLVuV6BNpXS+jk/wAAAIB8efzMiNQYnmhxMElS53dOSX07ghgvmKySVTP0S5Mm/xdlZ5Va3TpYr9AYJBWr2dBq+PjyFN5Tv8X9XtILkEOxm/zMsx2VOiu0fNSj5uFXwxYhm+ypLyCX0GACbs39Olb83lyqGwtIYwBiXW1jtlFdQ5AzbHjkfgONxC6Rgm+F6AAAAIAkqv1WdLA24/BwGf0CcGqX+ORKV1ymjU3x43WJH/0mR8cs4wDt1v76vkBj/Vrv8C6X7x4eU1ZG8TvqIq4eRJl2vk4wG5rHZ1ttNwQHDW1InsxRQYGaio4aYVEC2kJHrKA/CoMsPuNaC8ZeaTNFoyG6ne7uB0RyyuS0z49cq/YNAg== dhaugh@dhaugh-ld2
```
Comment (nwalsh-lnk) [inline tests/test_plugins/test_crypto_plugins/keys/dsa-key.pub]: Remove your user and host from these files and replace with something generic. I'm guessing these keys aren't for anything sensitive, but can you confirm that?

File: tests/test_plugins/test_crypto_plugins/test_ecb.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def test_non_existent_file():
+    """The scan should not throw exceptions on a non-existent file"""
+    plugin = ECBCipherCheck()
+    non_existent_files = ['----']
+    try:
+        plugin.run(non_existent_files)
+    except:
+        assert False
```
Comment (nwalsh-lnk) [inline tests/test_plugins/test_crypto_plugins/test_ecb.py]: Why is this here? if we want the test to fail on a plugin fail we could just remove `plugin.run` and let it throw an Exception in the test.

File: qark/plugins/crypto/ecb_cipher_usage.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                    if method_name == 'getInstance' and qualifier == 'Cipher' and re.search(r'.*/ECB/.*',
+                                                                                            encryption_type):
+                        description = "getInstance should not be called with ECB as the cipher mode, as it is insecure."
+                        self.issues.append(
+                            Issue(self.category, self.name, self.severity, description, file_object=filepath))
+                except Exception:
+                    continue
+        except Exception:
+            log.exception("Couldn't parse the java file: %s", filepath)
+
```
[the remark this answers is not in the document]
Comment (dhaugh) [reply qark/plugins/crypto/ecb_cipher_usage.py]: I tried to recreate that, and passing Cipher.getInstance an empty string throws an exception

File: qark/plugins/utils.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+def run_regex(filename, rex):
+    """
+    Read a file line by line, run a regular expression against the content and return list of things that require inspection
+    """
+    things_to_inspect = []
+    try:
+        with open(filename) as f:
+            content = f.readlines()
+            for y in content:
```
[the remark this answers is not in the document]
Comment (dhaugh) [reply qark/plugins/utils.py]: Copied verbatim from old qark. Not convinced that the amount of memory would actually lead to performance issues. Will change name

File: tests/test_plugins/test_crypto_plugins/keys/dsa-key.pub (hunk 1 above)
Comment (dhaugh) [reply tests/test_plugins/test_crypto_plugins/keys/dsa-key.pub]: Generated just for this purpose - removed the usernames and host

File: tests/test_plugins/test_crypto_plugins/test_ecb.py (hunk 2 above)
Comment (dhaugh) [reply tests/test_plugins/test_crypto_plugins/test_ecb.py]: This is to make sure that the plugin doesn't bubble up the exception

File: qark/plugins/crypto/ecb_cipher_usage.py [hunk 5]
```diff
@@ -32,23 +29,24 @@ def _process_file(self, filepath):
...
+                                                                                        encryption_type):
+                    description = "getInstance should not be called with ECB as the cipher mode, as it is insecure."
+                    self.issues.append(
+                        Issue(self.category, self.name, self.severity, description, file_object=filepath))
+            except Exception:
+                continue
+
     def run(self, files, apk_constants=None):
-        relevant_files = [file for file in files if os.path.splitext(file)[1] == '.java']
+        relevant_files = [file_path for file_path in files if os.path.splitext(file)[1] == '.java']
```
Comment (nwalsh-lnk) [inline qark/plugins/crypto/ecb_cipher_usage.py]: `os.path.splitext(file)` instead of `os.path.splitext(file_path)`

File: qark/plugins/crypto/ecb_cipher_usage.py [hunk 6]
```diff
@@ -32,23 +29,24 @@ def _process_file(self, filepath):
...
         for file in relevant_files:
```
Comment (nwalsh-lnk) [inline qark/plugins/crypto/ecb_cipher_usage.py]: Rename this as well

File: qark/plugins/crypto/setting_secure_random_seed.py [hunk 7]
```diff
@@ -9,27 +10,49 @@
...
 
         self.severity = Severity.WARNING
-        self.parser = Parser()
         self.tree = None
 
-    def run(self, file_object):
+    def _imports_secure_seed(self, tree):
+        """Checks if a tree imports java.security.SecureRandom, and returns True if the import exists"""
+        imports = tree.filter(javalang.tree.Import)
+        for path, curr_import in imports:
```
Comment (nwalsh-lnk) [inline qark/plugins/crypto/setting_secure_random_seed.py]: If you aren't using the path param, it should be `_` by convention.

File: qark/plugins/crypto/setting_secure_random_seed.py [hunk 8]
```diff
@@ -9,27 +10,49 @@
...
+        try:
+            tree = javalang.parse.parse(body)
+        except Exception:
+            log.exception("Couldn't parse the java file: %s", filepath)
+
+        if not self._imports_secure_seed(tree):  # doesn't import the insecure function
+            return
+
+        method_invocations = tree.filter(javalang.tree.MethodInvocation)
+        for node_path, method_invocation_node in method_invocations:
```
Comment (nwalsh-lnk) [inline qark/plugins/crypto/setting_secure_random_seed.py]: `node_path` to `_`

File: tests/test_plugins/test_crypto_plugins/test_crypto_plugins.py [hunk 9]
```diff
@@ -38,13 +38,13 @@ def test_packaged_private_keys():
     plugin = PackagedPrivateKeys()
     assert len(plugin.issues) is 0
```
Comment (nwalsh-lnk) [inline tests/test_plugins/test_crypto_plugins/test_crypto_plugins.py]: should be `==`

File: qark/plugins/crypto/ecb_cipher_usage.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+
+class ECBCipherCheck(BasePlugin):
+    def __init__(self):
+
+        BasePlugin.__init__(self, category="crypto", name="ECB Cipher Usage",
+                            description=("ECB mode is an insecure encryption technique and prone to data leakage"))
+
+        self.severity = Severity.VULNERABILITY
+        self.tree = None
```
Comment (nwalsh-lnk) [inline qark/plugins/crypto/ecb_cipher_usage.py]: This isn't used.

File: qark/plugins/crypto/setting_secure_random_seed.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    INSECURE_FUNCTIONS = ("setSeed", "generateSeed")
+
+    def __init__(self):
+
+        BasePlugin.__init__(self, category="crypto", name="Random number generator is seeded with SecureSeed",
+                            description=("Specifying a fixed seed will cause a predictable sequence of numbers. "
+                                         "This may be useful for testing, but not for secure use"))
+
+        self.severity = Severity.WARNING
+        self.tree = None
```
Comment (nwalsh-lnk) [inline qark/plugins/crypto/setting_secure_random_seed.py]: This isn't used.

File: qark/plugins/crypto/setting_secure_random_seed.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+    def _imports_secure_seed(self, tree):
+        """Checks if a tree imports java.security.SecureRandom, and returns True if the import exists"""
+        imports = tree.filter(javalang.tree.Import)
+        for path, curr_import in imports:
```
Comment (nwalsh-lnk) [inline qark/plugins/crypto/setting_secure_random_seed.py]: Change `path` to `_` if it is not being used, by convention.

File: qark/plugins/crypto/ecb_cipher_usage.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        pass
+        try:
+            with open(filepath, 'r') as f:
+                body = f.read()
+        except Exception:
+            log.exception("Unable to read file")
+            return
+
+        try:
+            tree = javalang.parse.parse(body)
```
[the remark this answers is not in the document]
Comment (dhaugh) [reply qark/plugins/crypto/ecb_cipher_usage.py]: addressed this - not sure why it's still showing up

File: qark/plugins/crypto/ecb_cipher_usage.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+                if method_name == 'getInstance' and qualifier == 'Cipher' and re.search(r'.*/ECB/.*',
+                                                                                        encryption_type):
+                    description = "getInstance should not be called with ECB as the cipher mode, as it is insecure."
+                    self.issues.append(
+                        Issue(self.category, self.name, self.severity, description, file_object=filepath))
+            except Exception:
+                continue
+
+    def run(self, files, apk_constants=None):
+        relevant_files = [file_path for file_path in files if os.path.splitext(file_path)[1] == '.java']
```
Comment (nwalsh-lnk) [inline qark/plugins/crypto/ecb_cipher_usage.py]: Not going to block merge but this should be a generator. I've realized there are a lot of places for generators in the new QARK code.

File: qark/plugins/crypto/packaged_private_keys.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    def __init__(self):
+        BasePlugin.__init__(self, category="crypto", name="Encryption keys are packaged with the application")
+
+        self.severity = Severity.VULNERABILITY
+
+    def run(self, files, apk_constants=None):
+        for file_path in files:
+            for regex in PackagedPrivateKeys.PRIVATE_KEY_REGEXES:
+                if run_regex(file_path, regex):
+                    log.info("It appears there is a private key embedded in your application: %s", file_path)
```
Comment (nwalsh-lnk) [inline qark/plugins/crypto/packaged_private_keys.py]: If we log something like this, make it debug. In v2 we shouldn't be continually outputting. The final report should have this same information anyways.

File: qark/plugins/crypto/setting_secure_random_seed.py [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+        BasePlugin.__init__(self, category="crypto", name="Random number generator is seeded with SecureSeed",
+                            description=("Specifying a fixed seed will cause a predictable sequence of numbers. "
+                                         "This may be useful for testing, but not for secure use"))
+
+        self.severity = Severity.WARNING
+
+    def _imports_secure_seed(self, tree):
+        """Checks if a tree imports java.security.SecureRandom, and returns True if the import exists"""
+        imports = tree.filter(javalang.tree.Import)
```
Comment (nwalsh-lnk) [inline qark/plugins/crypto/setting_secure_random_seed.py]: This should just be `tree.imports`. This data is already available and should save walking the entire tree.
